crates/project_model/src/workspace.rs

```diff
@@ -812,3 +844,40 @@ fn add_dep(graph: &mut CrateGraph, from: CrateId, name: CrateName, to: CrateId)
         log::error!("{}", err)
     }
 }
+
+/// Recreates the compile-time environment variables that Cargo sets.
+///
+/// Should be synced with
+/// <https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates>
+///
+/// FIXME: ask Cargo to provide this data instead of re-deriving.
+fn inject_cargo_env(package: &PackageData, env: &mut Env) {
+    // FIXME: Missing variables:
+    // CARGO_BIN_NAME, CARGO_BIN_EXE_<name>
+
+    let mut manifest_dir = package.manifest.clone();
+    manifest_dir.pop();
+    env.set("CARGO_MANIFEST_DIR".into(), manifest_dir.as_os_str().to_string_lossy().into_owned());
+
+    // Not always right, but works for common cases.
+    env.set("CARGO".into(), "cargo".into());
+
+    env.set("CARGO_PKG_VERSION".into(), package.version.to_string());
+    env.set("CARGO_PKG_VERSION_MAJOR".into(), package.version.major.to_string());
+    env.set("CARGO_PKG_VERSION_MINOR".into(), package.version.minor.to_string());
+    env.set("CARGO_PKG_VERSION_PATCH".into(), package.version.patch.to_string());
+    env.set("CARGO_PKG_VERSION_PRE".into(), package.version.pre.to_string());
+
+    env.set("CARGO_PKG_AUTHORS".into(), String::new());
+
+    env.set("CARGO_PKG_NAME".into(), package.name.clone());
+    // FIXME: This isn't really correct (a package can have many crates with different names), but
+    // it's better than leaving the variable unset.
+    env.set("CARGO_CRATE_NAME".into(), CrateName::normalize_dashes(&package.name).to_string());
+    env.set("CARGO_PKG_DESCRIPTION".into(), String::new());
+    env.set("CARGO_PKG_HOMEPAGE".into(), String::new());
+    env.set("CARGO_PKG_REPOSITORY".into(), String::new());
+    env.set("CARGO_PKG_LICENSE".into(), String::new());
```

Comment:
cc @jonas-schievink, this regresses https://github.com/rust-analyzer/rust-analyzer/pull/9579 a bit. 

It makes more sense to do this as a part of usual `to_crate_graph`, as opposed to when fetching build scripts, and it makes sense to use our own `CargoWorkspace` to derive this data, rather then the raw `cargo metadata`, but I didn't go the whole way of actually adding everything to `CargoWorkspace` (we probably want just a generic `Env` feild there, rather than a field for every bit of info). 

Ideally, we'd just make Cargo tell us all this vars directly, but it's unclear how one'd do it. 

Oh snap, I think I know a horrible hack to do this actually -- we already use `RUSC_WRAPPER` to short-cirtcuit the build, we can use the wrapper to also somehow echo the set of cargo vars it sees. Note: that something *could* be done doesn't mean that it *should* be done.

Reply:
This wrapper could theoretically also be used to create the full crate graph for non-cargo build systems instead of having to manually write `rust-project.json`.

Reply:
:thinking: that's interesting, yeah! I believe that's how CLion supports non-CMake-based projects.